ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/Genesis.hs

```diff
@@ -0,0 +1,590 @@
+
+-- | Given a list of candidate fragments and their associated peers, select the
+-- candidate to sync from. Return this fragment, the list of peers that are
+-- still in race to serve it, and the list of peers that are already being
+-- declined.
+selectTheCandidate ::
+  forall header peerInfo.
+  ( HasHeader header
+  ) =>
+  FetchDecisionPolicy header ->
+  -- | The current chain.
+  AnchoredFragment header ->
+  -- | The candidate fragments and their associated peers.
+  [(AnchoredFragment header, peerInfo)] ->
+  -- | The pair of: (a) a list of peers that we have decided are not right, eg.
+  -- because they presented us with a chain forking too deep, and (b) the
+  -- selected candidate that we choose to sync from and a list of peers that are
+  -- still in the race to serve that candidate.
+  WithDeclined
+    peerInfo
+    (Maybe (ChainSuffix header, [(ChainSuffix header, peerInfo)]))
+selectTheCandidate
+  FetchDecisionPolicy {compareCandidateChains, plausibleCandidateChain}
+  currentChain =
+    separateDeclinedAndStillInRace
+      -- Select the suffix up to the intersection with the current chain. This can
+      -- eliminate candidates that fork too deep.
+      . selectForkSuffixes currentChain
+      -- Filter to keep chains the consensus layer tells us are plausible.
+      . filterPlausibleCandidates plausibleCandidateChain currentChain
+    where
+      -- Write all of the declined peers, and find the longest candidate
+      -- fragment if there is any.
+      separateDeclinedAndStillInRace ::
+        [(FetchDecision (ChainSuffix header), peerInfo)] ->
+        WithDeclined peerInfo (Maybe (ChainSuffix header, [(ChainSuffix header, peerInfo)]))
+      separateDeclinedAndStillInRace decisions = do
+        let (declined, inRace) = partitionEithers
+              [ bimap ((,p)) ((,p)) d | (d, p) <- decisions ]
+        tell (DList.fromList declined)
+        case inRace of
+          [] -> pure Nothing
+          _ : _ -> do
+            let maxChainOn f c0 c1 = case compareCandidateChains (f c0) (f c1) of
+                  LT -> c1
+                  _  -> c0
+                -- maximumBy yields the last element in case of a tie while we
+                -- prefer the first one
+                chainSfx = fst $
+                  List.foldl1' (maxChainOn (getChainSuffix . fst)) inRace
+            pure $ Just (chainSfx, inRace)
+
+-- | Given _the_ candidate fragment to sync from, and a list of peers (with
+-- their corresponding candidate fragments), choose which peer to sync _the_
+-- candidate fragment from.
+--
+-- We first filter out all the peers that cannot even serve a reasonable batch
+-- of _the_ candidate fragment, and then we choose the first one according to
+-- the ordering passed as argument.
+--
+-- PRECONDITION: The set of peers must be included in the peer order queue.
+-- PRECONDITION: The given candidate fragments must not be empty.
+selectThePeer ::
```

Comment:
> This function will choose the longest fragment from one of the peers in the `PeersOrder`.  The fact that we will be downloading blocks from one peers must come from another place.  Is this because of `chain-sync` jumping?

I am not sure I understand your remark/question. Are you asking for the motivation for why it is done this way?

> Btw, does `dynamo` (which is used in various places in haddocs), mean the peer from which we are downloading headers, as opposed to the 'jumping' ones?

Exactly, "dynamo" is the term for that :+1: